src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -426,45 +426,53 @@ public static string SchemeToString(HttpScheme scheme)
             }
         }
 
-        public static bool IsValidHostHeader(string hostText)
+        public static void ValidateHostHeader(string hostText)
         {
-            // The spec allows empty values
-            if (string.IsNullOrEmpty(hostText))
+            // This is a string.IsNullOrEmpty test, but arranged to elmininate the
+            // bounds check from accessing the firstChar of the string
+            if (hostText is null || 0u >= (uint)hostText.Length)
             {
-                return true;
+                // The spec allows empty values
+                return;
             }
 
-            if (hostText[0] == '[')
+            var firstChar = hostText[0];
+            if (firstChar == '[')
             {
-                return IsValidIPv6Host(hostText);
+                // Tail call
+                ValidateIPv6Host(hostText);
             }
-
-            if (hostText[0] == ':')
+            else
             {
-                // Only a port
-                return false;
-            }
+                if (firstChar == ':')
+                {
+                    // Only a port
+                    BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
+                }
 
-            var i = 0;
-            for (; i < hostText.Length; i++)
-            {
-                if (!IsValidHostChar(hostText[i]))
+                // Enregister array
+                var hostCharValidity = HostCharValidity;
+                var i = 0;
+                for (; i < hostText.Length; i++)
                 {
-                    break;
+                    var ch = (int)hostText[i];
+                    // Bounds check and elimiate second bounds check
+                    if ((uint)ch >= (uint)hostCharValidity.Length || !hostCharValidity[ch])
```

Comment:
How is this faster than IsValidHostCharf? Does the second bounds check not get eliminated when `ch` is a char?

Reply:
No, was generating the asm while changing these and refining them. 

Perhaps `char` isn't treated as an unsigned value (e.g. `short` as `ushort` isn't cls complaint) so `ch < HostCharValidity.Length` could pass with a negative char?

Reply:
Also using a class level array `HostCharValidity` rather than a local array `hostCharValidity` can stop bounds checks being eliminated; and they recently made elimination more conservative https://github.com/dotnet/coreclr/pull/15756